lib/Partitioner/Partitioner.cpp

```diff
@@ -569,68 +581,390 @@ void Partitioner::doPartitioning(Function *F, NodeToFunctionMap &mapping) {
   }
 }
 
+void Partitioner::doFuncListPartitioning(llvm::StringRef funcName,
+                                         std::vector<Function *> funcs,
+                                         NodeToFunctionMap &mapping) {
+  // The dummy node.
+  DAGNodePtr DAGRoot = llvm::make_unique<DAGNode>();
+  DAGNodePtrVec nodes;
+  DAGRoot->logicalDevices = {0};
+  DAGRoot->name = funcName;
+  DAGRoot->module = module_;
+  DAGRoot->deviceIDs = {0};
+  DAGNode *root = DAGRoot.get();
+  llvm::DenseMap<Node *, Node *> currToNew;
+
+  // Clone nodes into target partition.
+  for (size_t i = 0, e = funcs.size(); i < e; i++) {
+    for (auto &N : funcs[i]->getNodes()) {
+      auto *clone = N.clone();
+      currToNew[&N] = clone;
+      mapping[&N]->addNode(clone);
+    }
+  }
+
+  // For any dependency that crosses a partition, add a placeholder and save
+  // node. Record the dependence in the function graph.
+  DeviceIDTy logicalID = 0;
+  std::unordered_map<NodeValue, Placeholder *> placeholders;
+  llvm::DenseMap<Function *, DAGNode *> funcDAG;
+  for (auto *subF : mapping.getPartitions()) {
+    if (funcDAG.find(subF) == funcDAG.end()) {
+      std::unique_ptr<DAGNode> subDAG = llvm::make_unique<DAGNode>();
+      subDAG->name = subF->getName();
+      subDAG->logicalDevices = {logicalID++};
+      subDAG->backendKind = mapping.getPartitionBackendKind(subF);
+      funcDAG[subF] = subDAG.get();
+      nodes.push_back(std::move(subDAG));
+    }
+
+    // Link subF to its parents.
+    std::set<Function *> parents;
+    for (auto &N : subF->getNodes()) {
+      for (int inp = 0, e = N.getNumInputs(); inp < e; inp++) {
+        auto input = N.getNthInput(inp);
+        if (isa<Storage>(input.getNode())) {
+          // It is possible that one input is the output of anther function.
+          Placeholder *ph = llvm::dyn_cast<Placeholder>(input.getNode());
```

Comment:
This code is very nested and difficult to follow:

https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

Reply:
@beicy The comment about nesting is not addressed yet.

Use `if (!isa<Storage>(input.getNode())) {continue;}` and later `if (!ph) {continue;}` as early exits to reduce nesting.

Reply:
Thanks! I will rewrite the logic as :  
```
if (isa<Storage>(input.getNode())) {
   if (isa<Constant>(input.getNode())) { continue;}
   //Handle Placeholder;
   continue;
}
```
The idea is, in one function, when we do the partition, we don't have to consider the Storage nodes. But if it possible that one Placeholder is the output of another function, then we need to handle this situation.